### ISSUE TITLE
Suggest errors.New over fmt.Errorf with single string parameter and no formatting

### ISSUE THREAD

**Issue by Jacalz** [dominikh/go-tools#1550]
I have seen a lot of code creating errors like `fmt.Errorf("something went wrong")` instead of `errors.New("something went wrong")`.  Not only is the latter faster, but I also think that a lot of the former cases are wrong in subtle ways. If you use `fmt.Errorf` with a single string and no formatting verbs, you either forgot to add formatting (kind of related to https://github.com/dominikh/go-tools/issues/1528, I guess) or you should have used a regular error instead. As such, I think it be flagged as a problem and suggest replacing the call. In the case where you don't want to wrap or format anything, I also think it is a lot clearer to see a regular `errors.New()` as I can easily know that it is a static error string.

**Comment by ccoVeille** [comment]
I assumed revive or another golangci-lint linter would detect it but it doesn't.

While it does detect fmt.Sprintf("whatever") is detected by gosimple S1039

**Comment by dominikh** [comment]
My concern with this check would be consistency with surrounding code. Should `fmt.Errorf("static string")` be permitted if somewhere else in the same file we have `fmt.Errorf("dynamic %d", n)`? How about somewhere else in the same package? I could see myself using `fmt.Errorf` if I was already using it and it means I didn't have to import `errors`.

**Comment by ccoVeille** [comment]
please note `govet` detects

```go
 fmt.Errorf("whatever", "whatever")
```
printf: fmt.Errorf call has arguments but no formatting directives (govet)

**Comment by ccoVeille** [comment]
> My concern with this check would be consistency with surrounding code. Should `fmt.Errorf("static string")` be permitted if somewhere else in the same file we have `fmt.Errorf("dynamic %d", n)`? How about somewhere else in the same package? I could see myself using `fmt.Errorf` if I was already using it and it means I didn't have to import `errors`.

I would recommend using `errors.New` anyway, as some linters could recommend using a sentinel error (nothing does this for now anyway, as far as I know)

**Comment by arp242** [comment]
> Not only is the latter faster

To put a number on that:

    func BenchmarkNew(b *testing.B) {
        for n := 0; n < b.N; n++ {
            _ = errors.New("Well hello, this is a static string that doesn't change")
        }
    }
    func BenchmarkErrorf(b *testing.B) {
        for n := 0; n < b.N; n++ {
            _ = fmt.Errorf("Well hello, this is a static string that doesn't change")
        }
    }

I get:

    % go test -bench=.
    goos: linux
    goarch: amd64
    pkg: tgo_20240529_FxuaSwr8
    cpu: Intel(R) Core(TM) i5-8350U CPU @ 1.70GHz
    BenchmarkNew-8          1000000000             0.2808 ns/op
    BenchmarkErrorf-8          9721765         139.3      ns/op

That was more than I had expected, but 139ns on a 2018 laptop is very little, especially since error paths are not often hot paths.

IMHO performance is not important here except in very special and rare cases.

---

With #1542 using a variable will get flagged; that can catch real errors – I have run into "accidental format string" with that, but never really with fixed strings.

This is in the same category as replacing `fmt.Errorf("%d", n)` with `strconv.Itoa()` (#1036): while there are real objective differences between the two, in a great many cases these differences don't matter, so practically speaking it's mostly a stylistic preference.

(The case for strconv is a bit stronger, because while *most* cases don't matter, it's *more* likely to lead to real measurable performance impact).

**Comment by ccoVeille** [comment]
So maybe something for perfsprintf linter ?

https://github.com/golangci/golangci-lint/pull/3714
 https://github.com/catenacyber/perfsprint/pull/1

**Comment by dominikh** [comment]
> `BenchmarkNew-8          1000000000             0.2808 ns/op`

The compiler is optimizing away your benchmark. errors.New gets inlined and the allocation gets optimized away. Try storing the result to a global variable instead of `_`.

**Comment by dominikh** [comment]
> I would recommend using errors.New anyway. Go "imports" errors in fmt.Errorf so it would change nothing I think.

My concern isn't the import graph, but the number of direct imports (and thus API made available) in a file. That is, I posit that having only one way of creating errors produces less mental burden than having two.

I'm not convinced by performance arguments either way, because the error path isn't the happy path. 

As for other linters: any linter that wants to suggest replacing calls to errors.New with exported error variables can do the same check for fmt.Errorf with a single argument.

**Comment by arp242** [comment]
Ah yeah, of course >_< It's ~31ns. Edited my message.

**Comment by Jacalz** [comment]
> I'm not convinced by performance arguments either way, because the error path isn't the happy path.

Agreed. That was not the reason for my suggestion related to this linter check. My main motivator is from a readability standpoint. I personally find `errors.New` a bit clearer and faster to parse when reading code as I can move on after just seeing `errors.N` instead of reading the whole error string and then trying to understand any formatting rules etc.

**Comment by ccoVeille** [comment]
Apparently, perfsprintf already reports it

https://github.com/catenacyber/perfsprint/issues/27#issuecomment-2139194813

I missed it, because I launched golangci-lint, but I missed that a local .golangci.yaml (very basic) file was present in the repository and hadn't perfsprintf enabled.

So, perfsprintf reports it already

**Comment by Jacalz** [comment]
Hmm. This seems to actually be triggered now with SA1006 as a call to `fmt.Errorf()` with a single string parameter results in a message like "printf-style function with dynamic format string and no further arguments should use print-style function instead" but that had me very confused initially. Perhaps the message should be adapted to be worded better in this case and we can close this issue afterwards?

**Comment by dominikh** [comment]
As of golang/go@c1e6e49d5d3f3fb927f1bfd1b453d8e7c906c6ac, performance is no longer an argument for this check. Furthermore, the commit message has this to say:

> For unformatted strings, it comes up periodically that there are
> more allocations using fmt.Errorf("x") compared to errors.New("x").
> People cite it as a reason to switch code using fmt.Errorf to
> use errors.New instead.
>
> [...]
> 
> On the other hand, I believe the position of the core Go team
> is that it is usually not worthwhile to make such a change. For example,
> in #52696, Russ wrote:
> 
>> Thanks for raising the issue, but please don't do this. Using
>> fmt.Errorf("foo") is completely fine, especially in a program where
>> all the errors are constructed with fmt.Errorf. Having to
>> mentally switch between two functions based on the argument
>> is unnecessary noise.

which means that isn't something we should flag.

As for

> This seems to actually be triggered now with SA1006 as a call to fmt.Errorf() with a single string parameter results in a message like "printf-style function with dynamic format string and no further arguments should use print-style function instead"

that check is about _dynamic_ string arguments, not constant ones. That has nothing to do with `errors.New` being simpler or faster, but the fact that dynamic strings might contain `%` signs, which would lead to mangled errors.
